fallbackNodePolyfills: false -> Module not found: Can't resolve 'path' in `next/future/image` (micromatch)

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 21.6.0: Wed Aug 10 14:28:23 PDT 2022; root:xnu-8020.141.5~2/RELEASE_ARM64_T6000
    Binaries:
      Node: 16.16.0
      npm: 8.15.1
      Yarn: 3.2.1
      pnpm: N/A
    Relevant packages:
      next: 12.3.0
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0

### What browser are you using? (if relevant)

Edge 105.0.1343.27 (Official build) (arm64)

### How are you deploying your application? (if relevant)

not relevant but, `next dev` / `next build` / `next start`

### Describe the Bug

12.3.0 added a new experimental config option called `fallbackNodePolyfills` to prevent polyfilling and getting feedback when you might use an internal node module inside your client.

https://nextjs.org/blog/next-12-3#other-improvements

This seems to clash with `next/future/image` using `micromatch` which uses node specific modules.

```
../../node_modules/next/dist/compiled/micromatch/index.js:22:3426
Module not found: Can't resolve 'path'

Import trace for requested module:
../../node_modules/next/dist/shared/lib/match-remote-pattern.js
../../node_modules/next/dist/client/future/image.js
../../node_modules/next/future/image.js
./src/components/SidebarLayout.tsx
./src/pages/docs/[...slug].tsx

https://nextjs.org/docs/messages/module-not-found
```

### Expected Behavior

`next/future/image` not throwing and working correctly with `fallbackNodePolyfills: false`

### Link to reproduction

https://stackblitz.com/edit/vercel-next-js-sckjm9?file=next.config.js

### To Reproduce

Just open the stackblitz, thats all to it.

```tsx
import Image from 'next/future/image';

/** Add your relevant code here for the issue to reproduce */
export default function Home() {
  return <Image src='https://raw.githubusercontent.com/discordjs/discord.js/main/.github/powered-by-vercel.svg' width={100} height={100} />;
}
```

```js
/** @type {import("next").NextConfig} */
module.exports = {
  reactStrictMode: true,
  images: {
    domains: ['raw.githubusercontent.com'],
  },
  experimental: {
    fallbackNodePolyfills: false,
  },
};
```

--- a/packages/next/build/webpack-config.ts
+++ b/packages/next/build/webpack-config.ts
@@ -1590,7 +1590,30 @@ export default async function getBaseWebpackConfig(
                     resolve: {
                       fallback:
                         config.experimental.fallbackNodePolyfills === false
-                          ? {}
+                          ? {
+                              assert: false,
+                              buffer: false,
+                              constants: false,
+                              crypto: false,
+                              domain: false,
+                              http: false,
+                              https: false,
+                              os: false,
+                              path: false,
+                              punycode: false,
+                              process: false,
+                              querystring: false,
+                              stream: false,
+                              string_decoder: false,
+                              sys: false,
+                              timers: false,
+                              tty: false,
+                              util: false,
+                              vm: false,
+                              zlib: false,
+                              events: false,
+                              setImmediate: false,
+                            }
                           : {
                               assert: require.resolve(
                                 'next/dist/compiled/assert'

--- a/scripts/next-with-deps.sh
+++ b/scripts/next-with-deps.sh
@@ -29,11 +29,11 @@ done
 
 if [ ! -z $HAS_CONFLICTING_DEP ] || [ ! -d "$PROJECT_DIR/node_modules" ];then
   cd $PROJECT_DIR
-  yarn install
+  pnpm install
   for dep in ${CONFLICTING_DEPS[@]};do 
     rm -rf node_modules/$dep
   done
 fi
 
 cd $START_DIR
-yarn next $@
+pnpm next $@

--- a/test/production/disable-fallback-polyfills/index.test.ts
+++ b/test/production/disable-fallback-polyfills/index.test.ts
@@ -4,6 +4,11 @@ import { NextInstance } from 'test/lib/next-modes/base'
 describe('Disable fallback polyfills', () => {
   let next: NextInstance
 
+  function getFirstLoadSize(output: string) {
+    const firstLoadRe = /○ \/\s*(\d*) k?B\s*(?<firstLoad>.*) kB/
+    return Number(output.match(firstLoadRe).groups.firstLoad)
+  }
+
   beforeAll(async () => {
     next = await createNext({
       files: {
@@ -19,21 +24,19 @@ describe('Disable fallback polyfills', () => {
           } 
         `,
       },
-      dependencies: {},
+      dependencies: {
+        axios: '0.27.2',
+      },
     })
     await next.stop()
   })
   afterAll(() => next.destroy())
 
   it('Fallback polyfills added by default', async () => {
-    const firstLoadJSSize = Number(
-      next.cliOutput.match(/○ \/\s{38}(\d*) kB\s{10}(?<firstLoad>\d*) kB/)
-        .groups.firstLoad
-    )
-    expect(firstLoadJSSize).not.toBeLessThan(200)
+    expect(getFirstLoadSize(next.cliOutput)).not.toBeLessThan(200)
   })
 
-  it('Build should fail, when fallback polyfilling is disabled', async () => {
+  it('Reduced bundle size when polyfills are disabled', async () => {
     await next.patchFile(
       'next.config.js',
       `module.exports = {
@@ -42,7 +45,27 @@ describe('Disable fallback polyfills', () => {
         }
       }`
     )
+    await next.start()
+    await next.stop()
 
-    await expect(next.start()).rejects.toThrow(/next build failed/)
+    expect(getFirstLoadSize(next.cliOutput)).toBeLessThan(200)
+  })
+
+  it('Pass build without error if non-polyfilled module is unreachable', async () => {
+    // `axios` uses `Buffer`, but it should be unreachable in the browser.
+    // https://github.com/axios/axios/blob/649d739288c8e2c55829ac60e2345a0f3439c730/lib/helpers/toFormData.js#L138
+    await next.patchFile(
+      'pages/index.js',
+      `import axios from 'axios'
+       import { useEffect } from 'react'
+
+       export default function Home() {
+         useEffect(() => {
+           axios.get('/api')
+         }, [])
+       }`
+    )
+
+    await expect(next.start()).not.toReject()
   })
 })

--- a/test/production/disable-fallback-polyfills/index.test.ts
+++ b/test/production/disable-fallback-polyfills/index.test.ts
@@ -63,6 +63,8 @@ describe('Disable fallback polyfills', () => {
          useEffect(() => {
            axios.get('/api')
          }, [])
+
+         return "hello world"
        }`
     )
 

